Sources/AsyncAlgorithms/MultiProducerSingleConsumerChannel/MultiProducerSingleConsumerChannel.swift

```diff
@@ -0,0 +1,489 @@
+public struct MultiProducerSingleConsumerChannelAlreadyFinishedError: Error {
+    @usableFromInline
+    init() {}
+}
+
+/// A multi producer single consumer channel.
+///
+/// The ``MultiProducerSingleConsumerChannel`` provides a ``MultiProducerSingleConsumerChannel/Source`` to
+/// send values to the channel. The source exposes the internal backpressure of the asynchronous sequence to the
+/// producer. Additionally, the source can be used from synchronous and asynchronous contexts.
+///
+///
+/// ## Using a MultiProducerSingleConsumerChannel
+///
+/// To use a ``MultiProducerSingleConsumerChannel`` you have to create a new channel with it's source first by calling
+/// the ``MultiProducerSingleConsumerChannel/makeChannel(of:throwing:BackpressureStrategy:)`` method.
+/// Afterwards, you can pass the source to the producer and the channel to the consumer.
+///
+/// ```
+/// let (channel, source) = MultiProducerSingleConsumerChannel<Int, Never>.makeChannel(
+///     backpressureStrategy: .watermark(low: 2, high: 4)
+/// )
+/// ```
+///
+/// ### Asynchronous producers
+///
+/// Values can be send to the source from asynchronous contexts using ``MultiProducerSingleConsumerChannel/Source/send(_:)-9b5do``
+/// and ``MultiProducerSingleConsumerChannel/Source/send(contentsOf:)-4myrz``. Backpressure results in calls
+/// to the `send` methods to be suspended. Once more elements should be produced the `send` methods will be resumed.
+///
+/// ```
+/// try await withThrowingTaskGroup(of: Void.self) { group in
+///     group.addTask {
+///         try await source.send(1)
+///         try await source.send(2)
+///         try await source.send(3)
+///     }
+///
+///     for await element in channel {
+///         print(element)
+///     }
+/// }
+/// ```
+///
+/// ### Synchronous producers
+///
+/// Values can also be send to the source from synchronous context. Backpressure is also exposed on the synchronous contexts; however,
+/// it is up to the caller to decide how to properly translate the backpressure to underlying producer e.g. by blocking the thread.
+///
+/// ## Finishing the source
+///
+/// To properly notify the consumer if the production of values has been finished the source's ``MultiProducerSingleConsumerChannel/Source/finish(throwing:)`` **must** be called.
+public struct MultiProducerSingleConsumerChannel<Element, Failure: Error>: AsyncSequence {
+    /// A private class to give the ``MultiProducerSingleConsumerChannel`` a deinit so we
+    /// can tell the producer when any potential consumer went away.
+    private final class _Backing: Sendable {
+        /// The underlying storage.
+        fileprivate let storage: _Storage
+
+        init(storage: _Storage) {
+            self.storage = storage
+        }
+
+        deinit {
+            storage.sequenceDeinitialized()
+        }
+    }
+
+    /// The backing storage.
+    private let backing: _Backing
+
+    @frozen
+    public struct ChannelAndStream: ~Copyable {
```

Comment:
Was it intended to call it `ChannelAndStream` while it is actually a container for `channel` and `source`?
Shouldn't it be called `ChannelAndSource`?